detectron2/modeling/meta_arch/fcos.py

```diff
@@ -96,77 +96,101 @@ def forward_training(self, images, features, predictions, gt_instances):
+        anchor_sizes = anchors.tensor[:, 2] - anchors.tensor[:, 0]  # (R, )
 
         lower_bound = anchor_sizes * 4
         lower_bound[: num_anchors_per_level[0]] = 0
         upper_bound = anchor_sizes * 8
         upper_bound[-num_anchors_per_level[-1] :] = float("inf")
 
-        matched_indices = []
-        for gt_per_image in gt_instances:
-            gt_centers = gt_per_image.gt_boxes.get_centers()  # Nx2
-            # FCOS with center sampling: anchor point must be close enough to gt center.
-            pairwise_match = (anchor_centers[:, None, :] - gt_centers[None, :, :]).abs_().max(
-                dim=2
-            ).values < self.center_sampling_radius * anchor_sizes[:, None]
-            pairwise_dist = pairwise_point_box_distance(anchor_centers, gt_per_image.gt_boxes)
-
-            # The original FCOS anchor matching rule: anchor point must be inside gt
-            pairwise_match &= pairwise_dist.min(dim=2).values > 0
-
-            # Multilevel anchor matching in FCOS: each anchor is only responsible
-            # for certain scale range.
-            pairwise_dist = pairwise_dist.max(dim=2).values
-            pairwise_match &= (pairwise_dist > lower_bound[:, None]) & (
-                pairwise_dist < upper_bound[:, None]
-            )
+        gt_centers = gt_boxes.get_centers()
+
+        # FCOS with center sampling: anchor point must be close enough to
+        # ground-truth box center.
+        center_dists = (anchor_centers[None, :, :] - gt_centers[:, None, :]).abs_()
+        sampling_regions = self.center_sampling_radius * anchor_sizes[None, :]
+
+        match_quality_matrix = center_dists.max(dim=2).values < sampling_regions
 
-            # Match the GT box with minimum area, if there are multiple GT matches
-            gt_areas = gt_per_image.gt_boxes.area()  # N
-            pairwise_match = pairwise_match.to(torch.float32) * (1e8 - gt_areas[None, :])
-            min_values, matched_idx = pairwise_match.max(dim=1)  # R, per-anchor match
-            matched_idx[min_values < 1e-5] = -1  # Unmatched anchors are assigned -1
+        pairwise_dist = pairwise_point_box_distance(anchor_centers, gt_boxes)
+        pairwise_dist = pairwise_dist.permute(1, 0, 2)  # (M, R, 4)
+
+        # The original FCOS anchor matching rule: anchor point must be inside GT.
+        match_quality_matrix &= pairwise_dist.min(dim=2).values > 0
+
+        # Multilevel anchor matching in FCOS: each anchor is only responsible
+        # for certain scale range.
+        pairwise_dist = pairwise_dist.max(dim=2).values
+        match_quality_matrix &= (pairwise_dist > lower_bound[None, :]) & (
+            pairwise_dist < upper_bound[None, :]
+        )
+        # Match the GT box with minimum area, if there are multiple GT matches.
+        gt_areas = gt_boxes.area()  # (M, )
 
-            matched_indices.append(matched_idx)
-        return matched_indices
+        # Get matches and their labels using match quality matrix.
```

Comment:
this line of comment seems unrelated

Reply:
Oops, an artifact of prior commit. Thanks for spotting!